sciplot/multiplot.hpp

```diff
@@ -0,0 +1,227 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+// C++ includes
+#include <vector>
+
+// sciplot includes
+#include <sciplot/constants.hpp>
+#include <sciplot/default.hpp>
+#include <sciplot/enums.hpp>
+#include <sciplot/palettes.hpp>
+#include <sciplot/plot.hpp>
+#include <sciplot/specs/axislabelspecs.hpp>
+#include <sciplot/specs/borderspecs.hpp>
+#include <sciplot/specs/gridspecs.hpp>
+#include <sciplot/specs/legendspecs.hpp>
+#include <sciplot/specs/linespecs.hpp>
+#include <sciplot/specs/plotspecs.hpp>
+#include <sciplot/specs/ticspecs.hpp>
+#include <sciplot/util.hpp>
+
+namespace sciplot
+{
+
+    using namespace internal;
+
+    /// The class used to define multipleplots in one canvas, show them in a pop-up window or save them to a file.
+    class multiplot : public plot
+    {
+    public:
+        /// Construct a default multiplot object
+        multiplot();
+
+        /// Set the palette of colors for the plot.
+        /// @param name Any palette name displayed in https://github.com/Gnuplotting/gnuplot-palettes, such as "viridis", "parula", "jet".
+        auto palette(const std::string &name) -> multiplot &;
+
+        /// Set the size of the plot (in unit of points, with 1 inch = 72 points).
+        auto size(std::size_t width, std::size_t height) -> multiplot &;
+
+        /// Set the layout of the multiplot in x- and y-direction.
+        auto layout(std::size_t rows, std::size_t columns) -> multiplot &;
+
+        /// Set the title of the multiplot.
+        auto title(const std::string &title) -> multiplot &;
+
+        /// Add a plot to the multiplot.
+        /// Note that if you're changing the plot afterwards, changes will NOT be recorded in the multiplot.
+        auto add(const plot &p) -> multiplot &;
+
+        /// Show the plot in a pop-up window.
+        auto show() -> void;
+
+        /// Save the plot in a file, with its extension defining the figure format.
+        /// The extension of the file name determines the format of the figure.
+        /// The supported figure formats are: `pdf`, `eps`, `svg`, `png`, and `jpeg`.
+        /// Thus, to save a plot in `png` format, choose a file name with a `.png`
+        /// file extension as in `fig.png`.
+        auto save(const std::string &filename) -> void;
+
+    protected:
+        /// Plot id derived from m_counter upon construction
+        auto id() const -> std::size_t;
+
+        /// Counter of how many plot / singleplot objects have been instanciated in the application
+        static std::size_t m_counter;
```

Comment:
How do we use this `static` counter? Is it possible not to use `static`?

Reply:
Well, some way the plot / multiplot instances have to be counted to not overwrite files and to connect script to data files. Callers don't care about that detail, so this way it is hidden from them. We could remove the id() method and just use m_id (yagni).

Reply:
I removed the id() function now and [made the counter private again](https://github.com/sciplot/sciplot/pull/10/commits/4d0aaf53100658a5c9649dfcb8f56a0350f14883). If we use [random temporary files](https://en.cppreference.com/w/cpp/io/c/tmpnam), we probably could get rid of it.

Reply:
OK, now I remember the use of this static counter. The disadvantage of using random temp files is that it is not clear which generated files correspond to the sequence of plot calls. I mean, if two plots are generated, their order indication in the file name may be useful to some users, for further inspecting the corresponding generated gnuplot scripts (check my comment here in the other issue you created #12 ). In the future, we could have different options on how files are generated, name schemes, and cleaning procedure for the generated gnuplot scripts.

Reply:
Let's take the discussion about this to #12